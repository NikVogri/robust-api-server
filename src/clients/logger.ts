import { Request, Response } from 'express';
import { Service } from 'typedi';
import winston from 'winston';
import { AppError } from '../error/AppError';

@Service()
export class Logger {
    private logger: winston.Logger;

    private custom = {
        levels: {
            error: 0,
            info: 2,
            http: 3,
        },
        colors: {
            error: 'red',
            info: 'yellow',
            http: 'gray',
        },
    };

    constructor() {
        winston.addColors(this.custom.colors);
        winston.remove(winston.transports.Console);

        this.logger = winston.createLogger({
            transports: this.getTransports(),
            levels: this.custom.levels,
        });
    }

    private getTransports = (): winston.transports.ConsoleTransportInstance[] => {
        return [
            // Logs appear in STDOUT
            new winston.transports.Console({
                level: 'http',
                format: winston.format.combine(winston.format.simple(), winston.format.colorize({ all: true })),
            }),
            new winston.transports.Console({
                level: 'info',
                format: winston.format.combine(winston.format.simple(), winston.format.colorize({ all: true })),
            }),
            new winston.transports.Console({
                level: 'error',
                format: winston.format.combine(
                    // winston.format.timestamp(), TODO: add timestamp
                    // TODO: make sure error only gets logged once! remove default transports for error?
                    winston.format.json(),
                    winston.format.colorize({ all: true }),
                ),
            }),
        ];
    };

    error(message: string, error: Error, metadata?: Record<string, unknown>) {
        this.logger.error(message, AppError.toStringifiableObject(error), metadata);
    }

    info(message: string, metadata?: Record<string, unknown>) {
        this.logger.info(message, metadata);
    }

    http(req: Request, res: Response, error?: AppError) {
        const log = this.buildHttpLog(req, res, error);
        this.logger.http(log);
    }

    private buildHttpLog(req: Request, res: Response, error?: AppError): string {
        return JSON.stringify({
            success: res.statusCode.toString().startsWith('2'), // success status codes are 2xx
            url: req.url,
            originalUrl: req.originalUrl,
            path: `${req.method} ${req.url}`,
            // TODO: Can also add auth fields in the future
            // user: {}
            client: {
                ip: req.ip,
                hostname: req.hostname,
                version: req.httpVersion,
                userAgent: req.headers['user-agent'],
            },
            req: {
                query: req.query,
                body: req.body ? Object.keys(req.body) : undefined,
                // do not log request inputs as they might be too large or contain GDPR data
            },
            res: {
                statusCode: res.statusCode,
                // do not log resposes as they might be too large or contain GDPR data
            },
            error: {
                message: error?.message,
                stack: error?.stack,
                thrown: error?.thrownError ? AppError.toStringifiableObject(error.thrownError) : undefined,
            },
        });
    }
}
