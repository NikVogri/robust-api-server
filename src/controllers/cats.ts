import { Service } from 'typedi';
import { CatsService } from '../services/cats';
import { Request, Response } from 'express';

@Service()
export class CatsController {
    constructor(private catsService: CatsService) {}

    async getAll(_: Request, res: Response) {
        const cats = await this.catsService.getAll();
        res.status(200).send(cats);
    }

    async getById(req: Request, res: Response) {
        const id = req.params.id;
        const cat = await this.catsService.getById(id);

        res.status(200).send(cat);
    }
}
