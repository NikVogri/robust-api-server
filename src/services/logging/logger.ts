import winston from 'winston';
import { AppError } from '../../error/AppError';
import { singleton } from 'tsyringe';
import { Request, Response } from '../../types';

@singleton()
export class Logger {
    public readonly winston: winston.Logger;

    private readonly custom = {
        levels: {
            error: 0,
            info: 1,
            http: 2,
            verbose: 3,
        },
        colors: {
            error: 'red',
            info: 'yellow',
            http: 'gray',
            verbose: 'gray',
        },
    };

    constructor() {
        winston.addColors(this.custom.colors);
        winston.remove(winston.transports.Console);

        this.winston = winston.createLogger({
            transports: this.getTransports(),
            levels: this.custom.levels,
        });
    }

    private getTransports = (): winston.transports.ConsoleTransportInstance[] => {
        return [
            // Logs appear in STDOUT
            new winston.transports.Console({
                level: 'verbose',
                format: winston.format.combine(
                    winston.format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss' }),
                    winston.format.printf(
                        (i) => `${i.level} ${i.timestamp}: ${i.message} ${JSON.stringify(i.stack) ?? ''}`,
                    ),
                    winston.format.colorize({ all: true }),
                ),
            }),
        ];
    };

    error(message: string, error: Error) {
        this.winston.error(message, AppError.toStringifiableObject(error));
    }

    info(message: string) {
        this.winston.info(message);
    }

    http(req: Request, res: Response, error?: Error) {
        const log = this.buildHttpLog(req, res, error);
        this.winston.http(log);
    }

    private buildHttpLog(req: Request, res: Response, error?: unknown): string {
        return JSON.stringify({
            success: res.statusCode.toString().startsWith('2'), // success status codes are 2xx
            originalUrl: req.originalUrl,
            path: `${req.method} ${req.url}`,
            // TODO: Can also add auth fields in the future
            // user: {}
            client: {
                ip: req.ip,
                hostname: req.hostname,
                version: req.httpVersion,
                userAgent: req.headers['user-agent'],
            },
            req: {
                query: Object.keys(req.query).length > 0 ? req.query : undefined,
                params: Object.keys(req.params).length > 0 ? req.params : undefined,
                body: Object.keys(req.body).length > 0 ? Object.keys(req.body) : undefined,
                // do not log request inputs as they might be too large or contain GDPR data
            },
            res: {
                statusCode: res.statusCode,
                // do not log resposes as they might be too large or contain GDPR data
            },
            error: this.buildErrorObject(error),
        });
    }

    private buildErrorObject(error?: unknown) {
        if (!error || !(error instanceof Error)) return undefined;

        const output: Record<string, unknown> = {
            message: error?.message,
        };

        if (error instanceof AppError && error.thrownError) {
            output['thrown'] = AppError.toStringifiableObject(error.thrownError);
        }

        return output;
    }
}
