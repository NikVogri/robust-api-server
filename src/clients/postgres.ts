import { Pool } from 'pg';
import { Logger } from './logger';
import { singleton } from 'tsyringe';

@singleton()
export class Postgres {
    readonly pool: Pool;

    constructor(private logger: Logger) {
        this.pool = new Pool({
            connectionString: process.env.PG_DB_CONNECTION_STR,
        });
    }

    async testConnection(): Promise<void> {
        await this.pool.query('SELECT 1 + 1;');
        this.logger.info('Successfully connected to Postgres');
    }

    async getClient() {
        return await this.pool.connect();
    }
}
