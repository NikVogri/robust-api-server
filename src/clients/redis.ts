import { Redis as IORedis } from 'ioredis';
import { Service } from 'typedi';
import { Logger } from './logger';

@Service()
export class Redis {
    client: IORedis;

    constructor(private logger: Logger) {
        this.client = new IORedis({
            maxRetriesPerRequest: null,
        });

        logger.info('Successfully connected to Redis');
    }
}
