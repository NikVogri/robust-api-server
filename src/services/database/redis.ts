import { Redis as IORedis } from 'ioredis';
import { Logger } from '../logging/logger';
import { singleton } from 'tsyringe';

@singleton()
export class Redis {
    readonly client: IORedis;

    constructor(private logger: Logger) {
        this.client = new IORedis(process.env.REDIS_DB_CONNECTION_STR!, {
            maxRetriesPerRequest: null,
        });
    }

    async testConnection(): Promise<void> {
        await this.client.ping();
        this.logger.info('Successfully connected to Redis');
    }
}
