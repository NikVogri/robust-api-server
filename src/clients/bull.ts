import { Queue, Worker } from 'bullmq';
import { Redis } from './redis';
import { catPetterWorker } from '../workers/catPetterWorker';
import { Logger } from './logger';
import { singleton } from 'tsyringe';

type QueueName = 'catPetterQueue';
type BullQueues = Record<QueueName, Queue<unknown, unknown, string>>;

@singleton()
export class Bull {
    readonly queues: BullQueues;

    constructor(
        private redis: Redis,
        private logger: Logger,
    ) {
        this.queues = this.createQueues();
        this.createWorkers();

        this.logger.info('Successfully initiated Bull queues and workers');
    }

    private createQueues(): BullQueues {
        return {
            catPetterQueue: new Queue('catPetterQueue', { connection: this.redis.client }),
        };
    }

    private createWorkers() {
        // In most cases workers should be running on seperate threads than the one the API is on
        new Worker('catPetterQueue', catPetterWorker, { connection: this.redis.client, concurrency: 30 });
    }
}
