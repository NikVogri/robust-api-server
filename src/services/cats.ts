import { Cat } from '../models/types';
import { AppError } from '../error/AppError';
import { Bull } from '../clients/bull';
import { CatsModel } from '../models/cats';
import { injectable } from 'tsyringe';
import { CreateCatDto } from '../routes/cats.dto';

@injectable()
export class CatsService {
    constructor(
        private catsModel: CatsModel,
        private bull: Bull,
    ) {}

    async getAll(): Promise<Cat[]> {
        const cats = await this.catsModel.getAll();
        return cats;
    }

    async getById(id: string): Promise<Cat> {
        const cat = await this.catsModel.getById(id);

        if (!cat) {
            throw new AppError({
                message: 'Cat does not exist :(',
                statusCode: 404,
            });
        }

        // don't forget to pet the cat
        await this.bull.queues.catPetterQueue.add('pet-cat', { name: cat.name });

        return cat;
    }

    async createCat(createCatDto: CreateCatDto): Promise<Cat> {
        const cat = await this.catsModel.create(createCatDto);
        return cat;
    }
}
