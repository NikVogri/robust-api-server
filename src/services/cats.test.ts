import { describe, expect, it, vi } from 'vitest';
import { CatsService } from './cats';
import { CatsRepository } from '../repositories/cats';
import { Bull } from './async/bull';
import { Postgres } from './database/postgres';
import { Logger } from './logging/logger';
import { Redis } from './database/redis';
import { CatDBItem } from '../models';
import { Job } from 'bullmq';
import { AppError } from '../error/AppError';

vi.mock('../clients/logger');
vi.mock('../clients/postgres');
vi.mock('../clients/redis');
vi.mock('../clients/bull', () => {
    return {
        Bull: vi.fn(() => ({
            queues: {
                catPetterQueue: {
                    add: vi.fn(),
                },
            },
        })),
    };
});
vi.mock('../repositories/cats');

describe('Test CatsService', () => {
    const logger = new Logger();

    const pg = new Postgres(logger);
    const redis = new Redis(logger);

    const bull = new Bull(redis, logger);
    const catsRepository = new CatsRepository(pg);

    const catsService = new CatsService(catsRepository, bull);

    it('should fetch all the cats', async () => {
        expect.assertions(2);

        vi.spyOn(catsRepository, 'getAll').mockImplementation(async (...args) => {
            expect(args).toMatchSnapshot();
            return [];
        });

        const res = await catsService.getAll();
        expect(res).toMatchSnapshot();
    });

    it('should fetch a single cat and pet it', async () => {
        expect.assertions(3);

        vi.spyOn(catsRepository, 'getById').mockImplementationOnce(async (...args) => {
            expect(args).toMatchSnapshot();
            return Promise.resolve({ name: 'Jake' } as CatDBItem);
        });

        vi.spyOn(bull.queues.catPetterQueue, 'add').mockImplementationOnce(async (...args) => {
            expect(args).toMatchSnapshot();
            return Promise.resolve({}) as Promise<Job<unknown, unknown>>;
        });

        const res = await catsService.getById('1');
        expect(res).toMatchSnapshot();
    });

    it("should fail fetching a single cat because it doesn't exist", async () => {
        expect.assertions(4);

        vi.spyOn(catsRepository, 'getById').mockImplementationOnce(async (...args) => {
            expect(args).toMatchSnapshot();
            return Promise.resolve(undefined);
        });

        const catPetterQueueAdd = vi.spyOn(bull.queues.catPetterQueue, 'add');

        try {
            await catsService.getById('1');
        } catch (error) {
            expect(error).toBeInstanceOf(AppError);
            expect(error).toMatchSnapshot();
        }

        expect(catPetterQueueAdd).not.toBeCalled();
    });
});
