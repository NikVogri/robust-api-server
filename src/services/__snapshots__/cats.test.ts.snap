// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test CatsService > should fail fetching a single cat because it doesn't exist 1`] = `
[
  "1",
]
`;

exports[`Test CatsService > should fail fetching a single cat because it doesn't exist 2`] = `[Error: Cat does not exist :(]`;

exports[`Test CatsService > should fetch a single cat and pet it 1`] = `
[
  "1",
]
`;

exports[`Test CatsService > should fetch a single cat and pet it 2`] = `
[
  "pet-cat",
  {
    "name": "Jake",
  },
]
`;

exports[`Test CatsService > should fetch a single cat and pet it 3`] = `
{
  "name": "Jake",
}
`;

exports[`Test CatsService > should fetch all the cats 1`] = `[]`;

exports[`Test CatsService > should fetch all the cats 2`] = `[]`;
